cmd/object-handlers.go

```diff
@@ -547,19 +699,51 @@ func (api objectAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Req
 	})
 }
 
-// HeadObjectHandler - HEAD Object
-// -----------
-// The HEAD operation retrieves metadata from an object without returning the object itself.
-func (api objectAPIHandlers) HeadObjectHandler(w http.ResponseWriter, r *http.Request) {
-	ctx := newContext(r, w, "HeadObject")
+// getFilesFromZIPObject reads a partial stream of a zip file to build the zipindex.Files index
+func getFilesListFromZIPObject(ctx context.Context, objectAPI ObjectLayer, bucket, object string) (zipindex.Files, ObjectInfo, error) {
+	var size = 10 << 10
+	var objSize int64
+	for {
+		rs := &HTTPRangeSpec{IsSuffixLength: true, Start: int64(-size)}
+		gr, err := objectAPI.GetObjectNInfo(ctx, bucket, object, rs, nil, readLock, ObjectOptions{})
+		if err != nil {
+			return nil, ObjectInfo{}, err
+		}
+		b, err := stdioutil.ReadAll(gr)
+		if err != nil {
+			gr.Close()
+			return nil, ObjectInfo{}, err
+		}
+		gr.Close()
+		if size > len(b) {
+			size = len(b)
+		}
 
-	defer logger.AuditLog(ctx, w, r, mustGetClaimsFromToken(r))
+		// Calculate the object real size if encrypted
+		if _, ok := crypto.IsEncrypted(gr.ObjInfo.UserDefined); ok {
+			objSize, err = gr.ObjInfo.DecryptedSize()
+			if err != nil {
+				return nil, ObjectInfo{}, err
+			}
+		} else {
+			objSize = gr.ObjInfo.Size
+		}
 
-	objectAPI := api.ObjectAPI()
-	if objectAPI == nil {
-		writeErrorResponseHeadersOnly(w, errorCodes.ToAPIErr(ErrServerNotInitialized))
-		return
+		files, err := zipindex.ReadDir(b[len(b)-size:], objSize, nil)
+		if err == nil {
+			return files, gr.ObjInfo, nil
+		}
+		var terr zipindex.ErrNeedMoreData
+		if errors.As(err, &terr) {
+			size = int(terr.FromEnd)
```

Comment:
```suggestion
			size = int(terr.FromEnd)
			if size <=0 || size > 100 << 20 {
				return nil, ObjectInfo{}, errors.New("zip directory too large")
			}
```
This will avoid hostile OOM attacks.

Reply:
> This will avoid hostile OOM attacks.

Nice catch!